saleor/graphql/meta/resolvers.py

```diff
@@ -37,33 +39,40 @@ def resolve_object_with_metadata_type(instance: ModelWithMetadata):
-        product_models.DigitalContent: product_types.DigitalContent,
-        order_models.Fulfillment: order_types.Fulfillment,
-        giftcard_models.GiftCard: giftcard_types.GiftCard,
-        order_models.Order: order_types.Order,
-        invoice_models.Invoice: invoice_types.Invoice,
-        page_models.Page: page_types.Page,
-        page_models.PageType: page_types.PageType,
-        payment_models.Payment: payment_types.Payment,
-        product_models.Product: product_types.Product,
-        product_models.ProductType: product_types.ProductType,
-        product_models.ProductVariant: product_types.ProductVariant,
-        menu_models.Menu: menu_types.Menu,
-        menu_models.MenuItem: menu_types.MenuItem,
-        shipping_models.ShippingMethod: shipping_types.ShippingMethod,
-        shipping_models.ShippingZone: shipping_types.ShippingZone,
-        account_models.User: account_types.User,
-        warehouse_models.Warehouse: warehouse_types.Warehouse,
-        discount_models.Sale: discount_types.Sale,
-        discount_models.Voucher: discount_types.Voucher,
-    }
-    return MODEL_TO_TYPE_MAP.get(instance.__class__, None)
+    if isinstance(instance, ModelWithMetadata):
+        MODEL_TO_TYPE_MAP = {
+            app_models.App: app_types.App,
+            attribute_models.Attribute: attribute_types.Attribute,
+            product_models.Category: product_types.Category,
+            checkout_models.Checkout: checkout_types.Checkout,
+            product_models.Collection: product_types.Collection,
+            product_models.DigitalContent: product_types.DigitalContent,
+            order_models.Fulfillment: order_types.Fulfillment,
+            giftcard_models.GiftCard: giftcard_types.GiftCard,
+            order_models.Order: order_types.Order,
+            invoice_models.Invoice: invoice_types.Invoice,
+            page_models.Page: page_types.Page,
+            page_models.PageType: page_types.PageType,
+            payment_models.Payment: payment_types.Payment,
+            product_models.Product: product_types.Product,
+            product_models.ProductType: product_types.ProductType,
+            product_models.ProductVariant: product_types.ProductVariant,
+            menu_models.Menu: menu_types.Menu,
+            menu_models.MenuItem: menu_types.MenuItem,
+            shipping_models.ShippingMethod: shipping_types.ShippingMethod,
+            shipping_models.ShippingZone: shipping_types.ShippingZone,
+            account_models.User: account_types.User,
+            warehouse_models.Warehouse: warehouse_types.Warehouse,
+            discount_models.Sale: discount_types.Sale,
+            discount_models.Voucher: discount_types.Voucher,
+        }
+        return MODEL_TO_TYPE_MAP.get(instance.__class__, None), instance.pk
+
+    elif dataclasses.is_dataclass(instance):
+        DATACLASS_TO_TYPE_MAP = {
+            shipping_interface.ShippingMethodData: shipping_types.ShippingMethod
```

Comment:
You have added a logic here but it's not tested. Please test this case.

Reply:
On `main` this can happen for checkouts, it's an unusual use case as shipping permissions are needed and there's no test for this currently.

After #8737, `ShippingMethodData` will be used for order drafts as well, where it's easier to access the metadata as an admin.

Regardless of that, it would be a good idea to make a checkout query as an app with the required permissions as some 3rd party apps could use that.

Reply:
Added adequate test.